### PR TITLE
core(tsc): add type checking to dbw audits

### PR DESCRIPTION
**Pull request by brendankenny**
the only major cleanups are in the `no-vulnerable-libraries.js.js` (lots of mutations) and `event-helpers.js` (simplified so that the type checker could follow along).

The golden lhr changes are eliminating the duplicate vulnerability listings (because the original library objects are no longer mutated) and taking out the `label` entry on the event listeners since it appears to only be set, not read, and it's just a duplicate of the line/col info anyways.

### CODE REVIEW

File: lighthouse-core/audits/dobetterweb/no-vulnerable-libraries.js [hunk 1]
```diff
@@ -66,86 +70,103 @@ class NoVulnerableLibrariesAudit extends Audit {
...
-      if (lib.vulns.length > 0) {
-        lib.vulnCount = lib.vulns.length;
-        lib.highestSeverity = this.highestSeverity(lib.vulns).replace(/^\w/, l => l.toUpperCase());
-        totalVulns += lib.vulnCount;
-        lib.detectedLib = {};
-        lib.detectedLib.text = lib.name + '@' + lib.version;
-        lib.detectedLib.url = lib.pkgLink;
-        lib.detectedLib.type = 'link';
+    /** @type {Array<{highestSeverity: string, vulnCount: number, detectedLib: LH.Audit.DetailsRendererLinkDetailsJSON}>} */
+    const vulnerableResults = [];
```
Comment (patrickhulce) [inline lighthouse-core/audits/dobetterweb/no-vulnerable-libraries.js]: vulnerabilities? vulnerabilityResults?

File: lighthouse-core/audits/dobetterweb/password-inputs-can-be-pasted-into.js [hunk 2]
```diff
@@ -23,15 +23,19 @@ class PasswordInputsCanBePastedIntoAudit extends Audit {
...
    */
   static audit(artifacts) {
     const passwordInputsWithPreventedPaste = artifacts.PasswordInputsWithPreventedPaste;
 
-    const items = passwordInputsWithPreventedPaste.map(input => ({
-      node: {type: 'node', snippet: input.snippet},
-    }));
+    /** @type {Array<{node: LH.Audit.DetailsRendererNodeDetailsJSON}>} */
+    const items = [];
+    passwordInputsWithPreventedPaste.forEach(input => {
```
Comment (patrickhulce) [inline lighthouse-core/audits/dobetterweb/password-inputs-can-be-pasted-into.js]: 😢 TS really can't deal with a map + assertion?

File: typings/audit.d.ts [hunk 3]
```diff
@@ -7,7 +7,7 @@
 declare global {
   module LH.Audit {
     export interface Context {
-      options: Object; // audit options
+      options: Record<string, any>; // audit options
```
Comment (patrickhulce) [inline typings/audit.d.ts]: what's the difference between `Record<K, V>` and `Object<K, V>`

File: typings/audit.d.ts [hunk 4]
```diff
@@ -47,7 +47,7 @@ declare global {
       debugString?: string;
     }
 
-    // TODO: placeholder typedefs until Details are typed
+    // TODO: placeholders typedefs until Details are typed
```
Comment (patrickhulce) [inline typings/audit.d.ts]: revert? :)

File: typings/audit.d.ts [hunk 5]
```diff
@@ -57,27 +57,36 @@ declare global {
     export interface DetailsRendererDetailsJSON {
       type: 'table';
       headings: Array<Audit.Heading>;
-      items: Array<{[x: string]: string|number|DetailsRendererNodeDetailsJSON}>;
+      items: Array<{[x: string]: DetailsItem}>;
       summary?: DetailsRendererDetailsSummary;
     }
 
+    export type DetailsItem = string | number | DetailsRendererNodeDetailsJSON |
```
Comment (patrickhulce) [inline typings/audit.d.ts]: DetailsItemMemberChildInfo

File: lighthouse-core/test/results/sample_v2.json [hunk 6]
```diff
@@ -3941,26 +3904,8 @@
         ],
         "items": [
           {
-            "name": "jQuery",
```
Comment (paulirish) [inline lighthouse-core/test/results/sample_v2.json]: lgtm on nuking this data. it's pretty low-value anyway

File: lighthouse-core/audits/dobetterweb/no-vulnerable-libraries.js (hunk 1 above)
Comment (brendankenny) [reply lighthouse-core/audits/dobetterweb/no-vulnerable-libraries.js]: went with vulnerabilityResults :)

File: typings/audit.d.ts (hunk 5 above)
Comment (brendankenny) [reply typings/audit.d.ts]: > DetailsItemMemberChildInfo

...Summary

File: typings/audit.d.ts (hunk 4 above)
Comment (brendankenny) [reply typings/audit.d.ts]: done

File: lighthouse-core/audits/dobetterweb/password-inputs-can-be-pasted-into.js (hunk 2 above)
Comment (brendankenny) [reply lighthouse-core/audits/dobetterweb/password-inputs-can-be-pasted-into.js]: it loses the type `'node'` (it widens to type `string`) :( The heuristics are wonky on widening like that, but people do want very different things at times (even we would probably usually want `string` in this case if it weren't for wanting to tell the difference between details types).

It's basically this or put a cast in there in addition to the declared type above, I believe, so this seemed better.

File: typings/audit.d.ts (hunk 3 above)
Comment (brendankenny) [reply typings/audit.d.ts]: > what's the difference between Record<K, V> and Object<K, V>

`Object<K, V>` only works in jsdoc. It's not parameterized in typescript code

Comment (brendankenny) [reply typings/audit.d.ts]: ...we should probably come up with some sort of style rule for this vs `{[p: string]: any}` vs `Object<>` on the jsdoc side

**Pull request merged**